### DISCUSSION TITLE
Ideas to improve/fix/implement for next version 2.3.0

### DISCUSSION THREAD

**Discussion by nacho4d** [XCTestHTMLReport/XCTestHTMLReport#226] [Ideas]
I write some ideas of things I have noticed we need to work. More ideas, better ideas, just ideas are off course always welcome ;-]

## Outstanding issues
- Fix time accuracy issues: https://github.com/XCTestHTMLReport/XCTestHTMLReport/issues/177 and https://github.com/XCTestHTMLReport/XCTestHTMLReport/issues/185 (I can work on these)
- Xcode13 new feature issue: https://github.com/XCTestHTMLReport/XCTestHTMLReport/issues/217
- Check this issue: https://github.com/XCTestHTMLReport/XCTestHTMLReport/issues/220 (Maybe same as #203)
- Fix install.sh https://github.com/XCTestHTMLReport/XCTestHTMLReport/issues/163

## Improvements
- Consider moving some of the integration tests from ruby to swift. XCTestHTMLReport now has some basic support for integration tests and everything that is tested with nokogiri can be tested with NDHpple's XMLPath (at least in theory). I haven't investigated much but at least for tests we can stop relying on ruby.
- Move to swift-argument-parser! (Already in the works by Tyler)
- Move completely to Github Actions (Github actions are already being leveraged now we just need to stop Travis). 
    - Extend release action to update brew formula too. https://github.com/XCTestHTMLReport/homebrew-xchtmlreport
- Improve the Wiki pages. 

## Investigation

- SPM now supports resources so we could have the templates as text files rather than having them as static strings inside a swift file. Currently we release a binary and that is all. This is super convenient. If by moving to SPM resources users will need a binary + extra text files then maybe this change does not worth it. 
OR
- Investigate about [new extensible build tools (Swift 5.6)](https://github.com/apple/swift-evolution/blob/main/proposals/0303-swiftpm-extensible-build-tools.md)  ... maybe we can keep the static strings approach but without using ruby scripts.

**Comment by nacho4d** [comment]
@tylervick I share this with you. If you have more ideas feel free to add them here. :)

**Comment by tylervick** [comment]
> Extend release action to update brew formula too.

I wasn't aware that this exists! We can probably remove the rb file in the main repo and make a downstream action after release to version bump.

Alternatively, we may be able to contribute directly to [Homebrew/homebrew-core](https://github.com/Homebrew/homebrew-core). Homebrew/homebrew-core#26977 was previously attempted but the repo is much larger now than it was at that time.

**Comment by tylervick** [comment]
Potential improvement: A webserver option in addition to the CLI. I.e. The ability to pass in xcresult bundles (either via filesystem or http upload) and output the generated html on the fly. I'm not sure how wide the use case is, but it might make sense given that these html bundles would often be uploaded to a (static) server anyway.

**Reply by tylervick** [reply]
POC: https://github.com/tylervick/XCTestHTMLReport/pull/5/files

**Comment by tylervick** [comment]
> SPM now supports resources so we could have the templates as text files rather than having them as static strings inside a swift file

This is probably worth looking into! Or at the very least compressing these resources to reduce binary size a bit.

https://github.com/stencilproject/Stencil is another option. While it might be overkill for the existing CLI use case, it might tie in nicely with the http server idea I posted above.
